api/machine/machine.proto

```diff
@@ -73,16 +73,33 @@ service MachineService {
 // ApplyConfiguration describes a request to assert a new configuration upon a
 // node.
 message ApplyConfigurationRequest {
+  enum Mode {
+    REBOOT = 0;
+    AUTO = 1;
+    NO_REBOOT = 2;
+    STAGED = 3;
+  }
   bytes data = 1;
-  bool on_reboot = 2;
-  bool immediate = 3;
+  // replaced by mode
+  bool on_reboot = 2 [
+    (common.remove_deprecated_field) = "v0.16",
+    deprecated = true
+  ];
+  // replaced by mode
+  bool immediate = 3 [
+    (common.remove_deprecated_field) = "v0.16",
+    deprecated = true
+  ];
+  Mode mode = 4;
 }
 
 // ApplyConfigurationResponse describes the response to a configuration request.
 message ApplyConfiguration {
   common.Metadata metadata = 1;
   // Configuration validation warnings.
   repeated string warnings = 2;
+  // Human readable auto mode apply detection result.
+  string auto_apply_info = 3;
```

Comment:
should we make it a bit more structured for easier API consumption (and probably in the future with `--dry-run`) by separating two things:

* `Mode mode` which states which mode was actually chosen (passes through mode most of the time, but for AUTO mode it returns either reboot or no-reboot).
* `string message` which might be a generic human-readable message explaining the result of the apply configuration call?